sound/soc/sof/ipc4-topology.c

```diff
@@ -819,6 +837,133 @@ static void sof_ipc4_unprepare_copier_module(struct snd_sof_widget *swidget)
+		depth = 24;
+		break;
+	case SNDRV_PCM_FORMAT_S32_LE:
+		depth = 32;
+		break;
+	default:
+		dev_err(sdev->dev, "invalid pcm frame format %d\n", params_format(params));
+		return -EINVAL;
+	}
+
+	*bit_depth = depth;
+	*channel_count = params_channels(params);
+	*sample_rate = params_rate(params);
+
+	dev_dbg(sdev->dev, "%s: rate %d channels %d sample_valid_bytes %d\n",
+		__func__, *sample_rate, *channel_count, *bit_depth);
+
+	return 0;
+}
+
+static int snd_sof_get_hw_config_params(struct snd_sof_dev *sdev, struct snd_sof_dai *dai,
+					int *sample_rate, int *channel_count, int *bit_depth)
+{
+	struct snd_soc_tplg_hw_config *hw_config;
+	struct snd_sof_dai_link *slink;
+	bool dai_link_found = false;
+	bool hw_cfg_found = false;
+	int i;
+
+	/* get current hw_config from link */
+	list_for_each_entry(slink, &sdev->dai_link_list, list) {
+		if (!strcmp(slink->link->name, dai->name)) {
+			dai_link_found = true;
+			break;
+		}
+	}
+
+	if (!dai_link_found) {
+		dev_err(sdev->dev, "%s: no DAI link found for DAI %s\n", __func__, dai->name);
+		return -EINVAL;
+	}
+
+	for (i = 0; i < slink->num_hw_configs; i++) {
+		hw_config = &slink->hw_configs[i];
+		if (dai->current_config == le32_to_cpu(hw_config->id)) {
+			hw_cfg_found = true;
+			break;
+		}
+	}
+
+	if (!hw_cfg_found) {
+		dev_err(sdev->dev, "%s: no matching hw_config found for DAI %s\n", __func__,
+			dai->name);
+		return -EINVAL;
+	}
+
+	*bit_depth = le32_to_cpu(hw_config->tdm_slot_width);
+	*channel_count = le32_to_cpu(hw_config->tdm_slots);
+	*sample_rate = le32_to_cpu(hw_config->fsync_rate);
+
+	dev_dbg(sdev->dev, "%s: rate %d channels %d sample_valid_bytes %d\n",
+		__func__, *sample_rate, *channel_count, *bit_depth);
+
+	return 0;
+}
+
+static int snd_sof_get_nhlt_endpoint_data(struct snd_sof_dev *sdev, struct snd_sof_dai *dai,
+					  struct snd_pcm_hw_params *params, u32 dai_index,
+					  u32 linktype, u8 dir, u32 **dst, u32 *len)
+{
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+	struct nhlt_specific_cfg *cfg;
+	int sample_rate, channel_count;
+	int bit_depth, ret;
+	u32 nhlt_type;
+
+	/* convert to NHLT type */
+	switch (linktype) {
+	case SOF_DAI_INTEL_DMIC:
+		nhlt_type = NHLT_LINK_DMIC;
+		ret = snd_sof_get_pcm_hw_params(sdev, params, &sample_rate, &channel_count,
+						&bit_depth);
+		break;
+	case SOF_DAI_INTEL_SSP:
+		nhlt_type = NHLT_LINK_SSP;
+		ret = snd_sof_get_hw_config_params(sdev, dai, &sample_rate, &channel_count,
+						   &bit_depth);
+		break;
+	default:
+		return 0;
+	}
+
+	if (ret < 0)
+		return ret;
+
+	dev_dbg(sdev->dev, "%s: dai_index %d nhlt_type %d dir %d\n",
+		__func__, dai_index, nhlt_type, dir);
+
+	/* try to find matching blob against hw_config params */
+	cfg = intel_nhlt_get_endpoint_blob(sdev->dev, ipc4_data->nhlt, dai_index, nhlt_type,
+					   bit_depth, bit_depth, channel_count, sample_rate,
+					   dir, 0);
+
+	if (!cfg) {
+		dev_err(sdev->dev,
+			"no matching blob for bit_depth: %d channels: %d sample_rate: %d\n",
```

Comment:
Would be nice to have some consistence in how we print.
line 864: "%s: rate %d channels %d sample_valid_bytes %d\n"
here:       "bit_depth: %d channels: %d sample_rate: %d\n"

4800Hz, 24bits, 2 channels as a format to print in all places? Or something, but consistent?